src/table.rs

```diff
@@ -0,0 +1,2013 @@
+    /// if let Ok(entry) = table.find_entry(hasher(&1), |val| val.0 == 1) {
+    ///     entry.remove();
+    /// }
+    /// assert_eq!(table.find(hasher(&1), |val| val.0 == 1), None);
+    /// # }
+    /// # fn main() {
+    /// #     #[cfg(feature = "nightly")]
+    /// #     test()
+    /// # }
+    /// ```
+    pub fn find_entry(
+        &mut self,
+        hash: u64,
+        eq: impl FnMut(&T) -> bool,
+    ) -> Result<OccupiedEntry<'_, T, A>, AbsentEntry<'_, T, A>> {
+        match self.raw.find(hash, eq) {
+            Some(bucket) => Ok(OccupiedEntry {
+                hash,
+                bucket,
+                table: self,
+            }),
+            None => Err(AbsentEntry { table: self }),
+        }
+    }
+
+    /// Returns an `Entry` for an entry in the table with the given hash
+    /// and which satisfies the equality function passed.
+    ///
+    /// This can be used to remove the entry from the table, or insert a new
+    /// entry with the given hash if one doesn't already exist.
+    ///
+    /// This method will call `eq` for all entries with the given hash, but may
+    /// also call it for entries with a different hash. `eq` should only return
+    /// true for the desired entry, at which point the search is stopped.
+    ///
+    /// This method may grow the table in preparation for an insertion. Call
+    /// [`HashTable::find_entry`] if this is undesirable.
+    ///
+    /// `hasher` is called if entries need to be moved or copied to a new table.
+    /// This must return the same hash value that each entry was inserted with.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// # #[cfg(feature = "nightly")]
+    /// # fn test() {
+    /// use ahash::AHasher;
+    /// use hashbrown::hash_table::Entry;
+    /// use hashbrown::HashTable;
+    /// use std::hash::{BuildHasher, BuildHasherDefault};
+    ///
+    /// let mut table = HashTable::new();
+    /// let hasher = BuildHasherDefault::<AHasher>::default();
+    /// let hasher = |val: &_| hasher.hash_one(val);
+    /// table.insert_unchecked(hasher(&1), (1, "a"), |val| hasher(&val.0));
+    /// if let Entry::Occupied(entry) = table.entry(hasher(&1), |val| val.0 == 1, |val| hasher(&val.0))
+    /// {
+    ///     entry.remove();
+    /// }
+    /// if let Entry::Vacant(entry) = table.entry(hasher(&2), |val| val.0 == 2, |val| hasher(&val.0)) {
+    ///     entry.insert((2, "b"));
+    /// }
+    /// assert_eq!(table.find(hasher(&1), |val| val.0 == 1), None);
+    /// assert_eq!(table.find(hasher(&2), |val| val.0 == 2), Some(&(2, "b")));
+    /// # }
+    /// # fn main() {
+    /// #     #[cfg(feature = "nightly")]
+    /// #     test()
+    /// # }
+    /// ```
+    pub fn entry(
+        &mut self,
+        hash: u64,
+        eq: impl FnMut(&T) -> bool,
+        hasher: impl Fn(&T) -> u64,
+    ) -> Entry<'_, T, A> {
+        match self.raw.find_or_find_insert_slot(hash, eq, hasher) {
```

Comment:
I'm playing with HashTable and came across a small problem. Can we directly provide some version of the `find_or_find_insert_slot` (***checked insert***) function?
It's very annoying that every time I need to insert a value I have to use the entry syntax (which is not that cheap) or use `find + insert_unchecked`, which is slow.

It can be something like this one:
```rust
pub fn insert<V>(
    &mut self,
    hash: u64,
    value: T,
    mut eq: impl FnMut(&T, &T) -> bool,
    hasher: impl Fn(&T) -> u64,
    replace: impl FnOnce(&mut T, T) -> V,
) -> Option<V> {
    match self
        .raw
        .find_or_find_insert_slot(hash, |found_val| eq(found_val, &value), hasher)
    {
        Ok(bucket) => Some(replace(unsafe { &mut bucket.as_mut() }, value)),
        Err(slot) => {
            unsafe {
                self.raw.insert_in_slot(hash, slot, value);
            }
            None
        }
    }
}
```
Then it can be used like:
```rust

pub struct NewMap<K, V, S = DefaultHashBuilder, A: Allocator = Global> {
    pub(crate) hash_builder: S,
    pub(crate) table: HashTable<(K, V), A>,
}

impl<K, V, S, A> NewMap<K, V, S, A>
where
    K: Eq + core::hash::Hash,
    S: core::hash::BuildHasher,
    A: Allocator,
{
    pub fn insert(&mut self, k: K, v: V) -> Option<V> {
        let hash = make_hash::<K, S>(&self.hash_builder, &k);
        let hasher = make_hasher::<_, V, S>(&self.hash_builder);

        self.table.insert(
            hash,
            (k, v),
            |found, new| found.0 == new.0,
            hasher,
            |(_, val_ref), (_, val)| core::mem::replace(val_ref, val),
        )
    }
}
```

Reply:
`entry` already maps directly to `find_or_find_insert_slot`. You can then use `Entry::insert` to unconditionally overwrite an existing value, or `Entry::or_insert` to only insert a new value if an old one doesn't already exist.